File: server/scripts/harvestAllSeedCorpus.js
/**
 * Round 4: Re-harvest evidence from ALL accumulated seed result directories.
 *
 * Reads from:
 *   1. .claude/seed_results/        (original, 196 files)
 *   2. .claude/seed_results_deep/   (round 1 — if exists)
 *   3. .claude/seed_results_batch2/ (round 2 — if exists)
 *   4. .claude/seed_results_batch3/ (round 3 — if exists)
 *
 * Flattens all comments + danmaku, mines against keyword dictionary,
 * merges new evidence, runs coverage audit.
 *
 * Usage:
 *   node server/scripts/harvestAllSeedCorpus.js          # dry run
 *   HARVEST_WRITE=1 node server/scripts/harvestAllSeedCorpus.js  # merge
 */

import { readFile, readdir, writeFile } from 'node:fs/promises';
import { basename, dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = join(__dirname, '..', '..');

const SOURCE_DIRS = [
  '.claude/seed_results',
  '.claude/seed_results_deep',
  '.claude/seed_results_batch2',
  '.claude/seed_results_batch3',
];

import { readKeywordDictionary } from '../services/deepseekKeywordTrainer.js';
import { findLocalCorpusEvidenceEntries } from '../services/localCorpusEvidence.js';

async function loadResultsFromDir(relPath) {
  const absPath = join(PROJECT_ROOT, relPath);
  const results = [];
  try {
    const files = await readdir(absPath);
    const jsonFiles = files.filter(f => f.endsWith('.json'));
    for (const file of jsonFiles) {
      try {
        const data = JSON.parse(await readFile(join(absPath, file), 'utf8'));
        if (data.seed && Array.isArray(data.videos)) results.push(data);
      } catch { /* skip */ }
    }
  } catch { /* dir doesn't exist yet */ }
  return results;
}

function flattenAllResults(allResults, dirLabel) {
  const comments = [];
  let totalComments = 0;
  let totalDanmaku = 0;
  let newestTs = 0;
  let oldestTs = Infinity;

  for (const seed of allResults) {
    const seedName = seed.seed || 'unknown';
    for (const video of seed.videos || []) {
      if (video.error) continue;
      const bvid = video.bvid || 'unknown';
      const source = `Bilibili history-tag harvest (${dirLabel}): https://www.bilibili.com/video/${bvid}/ (seed: ${seedName})`;

      for (const c of video.commentMessages || []) {
        const message = String(c.message || '').trim();
        if (!message || message.length < 2) continue;
        const ts = Number(c.time || 0);
        if (ts > 0) { if (ts > newestTs) newestTs = ts; if (ts < oldestTs) oldestTs = ts; }
        comments.push({ message, platform: 'bilibili', source, uid: bvid, uname: '' });
        totalComments++;
      }

      for (const d of video.danmakuMessages || []) {
        const message = String(d.message || '').trim();
        if (!message || message.length < 2) continue;
        const ts = Number(d.time || 0);
        if (ts > 0) { if (ts > newestTs) newestTs = ts; if (ts < oldestTs) oldestTs = ts; }
        comments.push({ message, platform: 'bilibili', source, uid: bvid, uname: '' });
        totalDanmaku++;
      }
    }
  }

  return { comments, totalComments, totalDanmaku, newestTs, oldestTs };
}

async function main() {
  console.log('=== Round 4: Multi-Source Evidence Re-Harvest ===\n');

  // 1. Load all source directories
  const allResults = [];
  for (const dir of SOURCE_DIRS) {
    const results = await loadResultsFromDir(dir);
    if (results.length > 0) {
      console.log(`  ${dir}: ${results.length} seed files`);
      allResults.push(...results);
    } else {
      console.log(`  ${dir}: (empty or not found)`);
    }
  }
  console.log(`  Total: ${allResults.length} seed result files\n`);

  // 2. Flatten
  console.log('Flattening comments + danmaku...');
  const { comments, totalComments, totalDanmaku, newestTs, oldestTs } = flattenAllResults(allResults, 'multi-round');
  console.log(`  ${totalComments.toLocaleString()} comments`);
  console.log(`  ${totalDanmaku.toLocaleString()} danmaku`);
  console.log(`  ${comments.length.toLocaleString()} total messages`);
  if (oldestTs < Infinity) {
    console.log(`  Date range: ${new Date(oldestTs * 1000).toISOString()} → ${new Date(newestTs * 1000).toISOString()}`);
  }

  // 3. Load dictionary
  console.log('\nLoading keyword dictionary...');
  const dictionary = await readKeywordDictionary();
  console.log(`  ${dictionary.entries?.length || 0} terms`);

  // 4. Mine evidence
  console.log('\nMining evidence matches...');
  const options = {
    targetEvidence: 5,
    maxSamplesPerTerm: 5,
    requireCommentBackedEvidence: true,
  };
  const evidenceEntries = findLocalCorpusEvidenceEntries(dictionary, comments, options);
  console.log(`  ${evidenceEntries.length} terms matched with new evidence`);

  if (evidenceEntries.length === 0) {
    console.log('\nNo new evidence found. Dictionary already fully covered.');
    return;
  }

  const newSamples = evidenceEntries.reduce((sum, e) => sum + (e.evidenceSamples?.length || 0), 0);
  console.log(`  ${newSamples} new evidence samples\n`);

  // 5. By family
  const byFamily = {};
  for (const entry of evidenceEntries) {
    const f = entry.family || 'unknown';
    if (!byFamily[f]) byFamily[f] = [];
    byFamily[f].push(entry);
  }
  console.log('New evidence by family:');
  for (const [family, entries] of Object.entries(byFamily).sort((a, b) => b[1].length - a[1].length)) {
    console.log(`  ${family}: ${entries.length} terms`);
  }

  // 6. Sample matches
  console.log('\nSample new evidence:');
  for (const entry of evidenceEntries.slice(0, 10)) {
    console.log(`  [${entry.family}] ${entry.term}: +${entry.evidenceSamples?.length || 0} samples`);
    for (const sample of (entry.evidenceSamples || []).slice(0, 2)) {
      console.log(`    → "${sample.slice(0, 80)}${sample.length > 80 ? '...' : ''}"`);
    }
  }

  // 7. Save report
  const report = {
    harvestedAt: new Date().toISOString(),
    source: 'multi-round deep scrape (seed_results + seed_results_deep + batch2 + batch3)',
    sourceFiles: allResults.length,
    totalComments,
    totalDanmaku,
    totalMessages: comments.length,
    dateRange: oldestTs < Infinity ? {
      oldest: new Date(oldestTs * 1000).toISOString(),
      newest: new Date(newestTs * 1000).toISOString(),
    } : null,
    matchedTerms: evidenceEntries.length,
    newEvidenceSamples: newSamples,
    byFamily: Object.fromEntries(
      Object.entries(byFamily).map(([f, entries]) => [f, entries.length])
    ),
    entries: evidenceEntries.map(e => ({
      term: e.term,
      family: e.family,
      newSamples: e.evidenceSamples?.length || 0,
      samples: e.evidenceSamples || [],
    })),
  };

  const reportPath = join(__dirname, '..', 'data', 'seedCorpusHarvestReportR4.json');
  await writeFile(reportPath, JSON.stringify(report, null, 2), 'utf8');
  console.log(`\nReport saved: ${reportPath}`);

  // 8. Merge
  if (process.env.HARVEST_WRITE === '1') {
    console.log('\nMerging evidence into dictionary...');
    const { mergeEntriesIntoDictionary } = await import('../services/deepseekKeywordTrainer.js');
    const updated = await mergeEntriesIntoDictionary(evidenceEntries);
    console.log(`  Dictionary updated: ${updated.entries?.length || 0} total entries`);
  } else {
    console.log('\nDry run — set HARVEST_WRITE=1 to merge.');
  }
}

main().catch(e => { console.error('Fatal:', e); process.exit(1); });

File: python_backend/cli/harvest_all_seed_corpus.py
"""
Round 4: Re-harvest evidence from ALL accumulated seed result directories.

Python contract for server/scripts/harvestAllSeedCorpus.js.
Produces identical JSON output for parity comparison.

Usage:
  python -m python_backend.cli.harvest_all_seed_corpus [--dry-run]
  python -m python_backend.cli.harvest_all_seed_corpus --merge
  python -m python_backend.cli.harvest_all_seed_corpus --compare-js-report <path>
"""

import argparse
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent

SOURCE_DIRS = [
    ".claude/seed_results",
    ".claude/seed_results_deep",
    ".claude/seed_results_batch2",
    ".claude/seed_results_batch3",
]


def load_results_from_dir(rel_path: str) -> list:
    """Load all seed result JSON files from a directory."""
    abs_path = PROJECT_ROOT / rel_path
    results = []
    if not abs_path.is_dir():
        return results
    for f in sorted(abs_path.glob("*.json")):
        try:
            data = json.loads(f.read_text(encoding="utf-8"))
            if data.get("seed") and isinstance(data.get("videos"), list):
                results.append(data)
        except (json.JSONDecodeError, OSError):
            pass
    return results


def flatten_all_results(all_results: list, dir_label: str) -> dict:
    """Flatten seed results into a unified comments list."""
    comments = []
    total_comments = 0
    total_danmaku = 0
    newest_ts = 0
    oldest_ts = float("inf")

    for seed in all_results:
        seed_name = seed.get("seed", "unknown")
        for video in seed.get("videos", []):
            if video.get("error"):
                continue
            bvid = video.get("bvid", "unknown")
            source = (
                f"Bilibili history-tag harvest ({dir_label}): "
                f"https://www.bilibili.com/video/{bvid}/ (seed: {seed_name})"
            )

            for c in video.get("commentMessages", []):
                message = str(c.get("message", "")).strip()
                if not message or len(message) < 2:
                    continue
                ts = int(c.get("time", 0) or 0)
                if ts > 0:
                    newest_ts = max(newest_ts, ts)
                    oldest_ts = min(oldest_ts, ts)
                comments.append({
                    "message": message,
                    "platform": "bilibili",
                    "source": source,
                    "uid": bvid,
                    "uname": "",
                })
                total_comments += 1

            for d in video.get("danmakuMessages", []):
                message = str(d.get("message", "")).strip()
                if not message or len(message) < 2:
                    continue
                ts = int(d.get("time", 0) or 0)
                if ts > 0:
                    newest_ts = max(newest_ts, ts)
                    oldest_ts = min(oldest_ts, ts)
                comments.append({
                    "message": message,
                    "platform": "bilibili",
                    "source": source,
                    "uid": bvid,
                    "uname": "",
                })
                total_danmaku += 1

    return {
        "comments": comments,
        "totalComments": total_comments,
        "totalDanmaku": total_danmaku,
        "newestTs": newest_ts,
        "oldestTs": oldest_ts if oldest_ts < float("inf") else 0,
    }


def build_report(all_results, flattened, evidence_entries, new_samples):
    """Build the standard JSON report matching JS output shape."""
    by_family = {}
    for entry in evidence_entries:
        f = entry.get("family", "unknown")
        by_family.setdefault(f, []).append(entry)

    entries_report = []
    for e in evidence_entries:
        entries_report.append({
            "term": e.get("term", ""),
            "family": e.get("family", "unknown"),
            "evidenceCount": len(e.get("evidenceSamples", [])),
            "evidenceSamples": e.get("evidenceSamples", [])[:5],
        })

    return {
        "ok": True,
        "harvestedAt": datetime.now(timezone.utc).isoformat(),
        "source": "multi-round deep scrape (seed_results + seed_results_deep + batch2 + batch3)",
        "sourceFiles": len(all_results),
        "totalComments": flattened["totalComments"],
        "totalDanmaku": flattened["totalDanmaku"],
        "totalMessages": len(flattened["comments"]),
        "dateRange": {
            "oldest": datetime.fromtimestamp(flattened["oldestTs"], tz=timezone.utc).isoformat() if flattened["oldestTs"] else None,
            "newest": datetime.fromtimestamp(flattened["newestTs"], tz=timezone.utc).isoformat() if flattened["newestTs"] else None,
        },
        "matchedTerms": len(evidence_entries),
        "newEvidenceSamples": new_samples,
        "byFamily": {f: len(entries) for f, entries in sorted(by_family.items())},
        "entries": entries_report,
    }


def main():
    parser = argparse.ArgumentParser(description="Harvest evidence from all seed result directories")
    parser.add_argument("--dry-run", action="store_true", default=True,
                        help="Print report without merging (default)")
    parser.add_argument("--merge", action="store_true",
                        help="Merge evidence into dictionary")
    parser.add_argument("--compare-js-report", type=str,
                        help="Path to JS report JSON for parity comparison")
    args = parser.parse_args()

    # 1. Load all source directories
    all_results = []
    for d in SOURCE_DIRS:
        results = load_results_from_dir(d)
        if results:
            all_results.extend(results)

    if not all_results:
        print(json.dumps({"ok": True, "matchedTerms": 0, "note": "no seed results found"}, ensure_ascii=False))
        return

    # 2. Flatten
    flattened = flatten_all_results(all_results, "multi-round")

    # 3. Mine evidence via local_corpus_evidence CLI (Python-owned, call as subprocess)
    import subprocess
    import tempfile

    try:
        dictionary_path = PROJECT_ROOT / "server" / "data" / "deepseekKeywordDictionary.production.json"
        if not dictionary_path.exists():
            dictionary_path = PROJECT_ROOT / "server" / "data" / "deepseekKeywordDictionary.json"

        # Write flattened comments to temp file for the CLI
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False,
                                         encoding="utf-8") as tf:
            json.dump(flattened["comments"], tf)
            comments_path = tf.name

        # Write payload for local_corpus_evidence CLI
        payload = {
            "dictionaryPath": str(dictionary_path),
            "commentsPath": comments_path,
            "targetEvidence": 5,
            "maxSamplesPerTerm": 5,
            "requireCommentBackedEvidence": True,
        }
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False,
                                         encoding="utf-8") as pf:
            json.dump(payload, pf)
            payload_path = pf.name

        result = subprocess.run(
            [sys.executable, "-m", "python_backend.cli.local_corpus_evidence",
             "--payload", payload_path],
            capture_output=True, text=True, timeout=60, cwd=str(PROJECT_ROOT),
            env={**os.environ, "PYTHONIOENCODING": "utf-8"},
        )

        if result.returncode == 0 and result.stdout.strip():
            evidence_data = json.loads(result.stdout)
            evidence_entries = evidence_data.get("entries", [])
        else:
            evidence_entries = []
    except (subprocess.TimeoutExpired, OSError, json.JSONDecodeError):
        evidence_entries = []
    finally:
        # Clean up temp files even on timeout or error
        for p in (comments_path, payload_path):
            try:
                os.unlink(p)
            except OSError:
                pass

    new_samples = sum(len(e.get("evidenceSamples", [])) for e in evidence_entries)

    # 4. Build report
    report = build_report(all_results, flattened, evidence_entries, new_samples)

    # 5. Handle --compare-js-report
    if args.compare_js_report:
        try:
            js_report = json.loads(Path(args.compare_js_report).read_text(encoding="utf-8"))
            match = (
                js_report.get("matchedTerms") == report["matchedTerms"]
                and js_report.get("totalMessages") == report["totalMessages"]
            )
            report["comparison"] = {
                "jsMatchedTerms": js_report.get("matchedTerms"),
                "pyMatchedTerms": report["matchedTerms"],
                "jsTotalMessages": js_report.get("totalMessages"),
                "pyTotalMessages": report["totalMessages"],
                "parity": match,
            }
        except (OSError, json.JSONDecodeError):
            report["comparison"] = {"error": "could not read JS report"}

    print(json.dumps(report, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
